pkg/controllers/cloud/networkpolicy.go

```diff
@@ -864,24 +878,91 @@ func (a *appliedToSecurityGroup) updateRules(r *NetworkPolicyReconciler) error {
 		a.hasRules = false
 		return nil
 	}
+	irules, erules := a.computeRules(nps)
+	if irules == nil || erules == nil {
+		return nil
+	}
+	r.Log.V(1).Info("AppliedToSecurityGroup update rules", "Name", a.id,
+		"ingressRules", irules, "egressRules", erules)
+	ch := securitygroup.CloudSecurityGroup.UpdateSecurityGroupRules(&a.id, irules, erules)
+	go func() {
+		err := <-ch
+		r.cloudResponse <- &securityGroupStatus{sg: a, op: securityGroupOperationUpdateRules, err: err}
+	}()
+	return nil
+}
+
+// computeRules compute all the ingress and egress rules of an appliedTo group.
+func (a *appliedToSecurityGroup) computeRules(nps []interface{}) ([]*securitygroup.IngressRule, []*securitygroup.EgressRule) {
 	irules := make([]*securitygroup.IngressRule, 0)
 	erules := make([]*securitygroup.EgressRule, 0)
 	for _, i := range nps {
 		np := i.(*networkPolicy)
 		if !np.rulesReady {
-			return nil
+			return nil, nil
 		}
 		irules = append(irules, deepcopy.Copy(np.ingressRules).([]*securitygroup.IngressRule)...)
 		erules = append(erules, deepcopy.Copy(np.egressRules).([]*securitygroup.EgressRule)...)
 	}
-	irules, erules = deduplicateIngressRules(irules), deduplicateEgressRules(erules)
-	r.Log.V(1).Info("AppliedToSecurityGroup update rules", "Name", a.id,
-		"ingressRules", irules, "egressRules", erules)
-	ch := securitygroup.CloudSecurityGroup.UpdateSecurityGroupRules(&a.id, irules, erules)
-	go func() {
-		err := <-ch
-		r.cloudResponse <- &securityGroupStatus{sg: a, op: securityGroupOperationUpdateRules, err: err}
-	}()
+	return deduplicateIngressRules(irules), deduplicateEgressRules(erules)
+}
+
+// updateAddrGroupReference updates appliedTo group addrGroupRef and notifies removed addrGroups that rules referencing them is removed.
+func (a *appliedToSecurityGroup) updateAddrGroupReference(r *NetworkPolicyReconciler) error {
+	// get latest irules and erules
+	nps, err := r.networkPolicyIndexer.ByIndex(networkPolicyIndexerByAppliedToGrp, a.id.Name)
+	if err != nil {
+		return fmt.Errorf("unable to get networkPolicy with key %s from indexer: %w", a.id.Name, err)
+	}
+	irules, erules := a.computeRules(nps)
+
+	// combine irules and erules to get latest addrGroupRef.
+	references := make(map[string]bool)
+	for _, rule := range irules {
+		for _, sg := range rule.FromSecurityGroups {
+			references[sg.String()] = true
+		}
+	}
+	for _, rule := range erules {
+		for _, sg := range rule.ToSecurityGroups {
+			references[sg.String()] = true
+		}
+	}
+	// compute addrGroupRef removed from previous.
+	removedRef := make([]string, 0)
```

Comment:
removedRefs? and update comment accordingly

Reply:
Done